src/Interpreters/InterserverCredentials.cpp

```diff
@@ -0,0 +1,69 @@
+#include <Interpreters/InterserverCredentials.h>
+#include <common/logger_useful.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int NO_ELEMENTS_IN_CONFIG;
+}
+
+std::shared_ptr<SimpleInterserverCredentials>
+SimpleInterserverCredentials::make(const Poco::Util::AbstractConfiguration & config, const std::string root_tag)
+{
+    const auto user = config.getString(root_tag + ".user", "");
+    const auto password = config.getString(root_tag + ".password", "");
+
+    if (user.empty())
+        throw Exception("Configuration parameter interserver_http_credentials user can't be empty", ErrorCodes::NO_ELEMENTS_IN_CONFIG);
+
+    auto store_ = makeCredentialStore(user, password, config, root_tag);
+
+    return std::make_shared<SimpleInterserverCredentials>(user, password, store_);
+}
+
+SimpleInterserverCredentials::Store SimpleInterserverCredentials::makeCredentialStore(
+    const std::string current_user_,
+    const std::string current_password_,
+    const Poco::Util::AbstractConfiguration & config,
+    const std::string root_tag)
+{
+    Store store_;
+    store_.push_back({current_user_, current_password_});
+    if (config.has(root_tag + ".allow_empty") && config.getBool(root_tag + ".allow_empty"))
+    {
+        /// Allow empty credential to support migrating from no auth
+        store_.push_back({"", ""});
+    }
+
+
+    Poco::Util::AbstractConfiguration::Keys users;
+    config.keys(root_tag + ".users", users);
+    for (const auto & user : users)
+    {
+        LOG_DEBUG(&Poco::Logger::get("InterserverCredentials"), "Adding credential for {}", user);
+        const auto password = config.getString(root_tag + ".users." + user);
+        store_.push_back({user, password});
+    }
+
+    return store_;
+}
+
+std::pair<std::string, bool> SimpleInterserverCredentials::isValidUser(std::pair<std::string, std::string> check)
+{
+    bool valid = false;
+    for (const auto & pair : store)
+    {
+        if (pair == check)
+        {
+            valid = true;
+            break;
+        }
+    }
+    if (!valid)
+        return {"Incorrect user or password in HTTP Basic authentification", false};
```

Comment:
why not via throw? Ok I see `InterserverIOHTTPHandler::checkAuthentication` is done in that way (but maybe it should be updated? (to ensure that the result will be 100% verified)

Reply:
Can throw if you want.StatusOr is probably less overhead (it's called from http handler). It's there a reason why the handler should throw?

Reply:
This will avoid further issues if someone will forget to check the return value (and since this is authorization this looks pretty critical)

P.S. if you will change this, you will need to guard the upper caller to wrap the call into the try/catch (this can be done separately I guess)

Reply:
IsValid shouldl throw and checkAuth will catch